src/libos/src/net/socket/unix/stream/socket.rs

```diff
@@ -0,0 +1,207 @@
+    }
+
+    pub fn addr(&self) -> Addr {
+        self.addr.read().unwrap().clone()
+    }
+
+    pub fn end(&self) -> &SgxMutex<Option<Arc<EndPoint>>> {
+        &self.end
+    }
+
+    pub fn peer_addr(&self) -> Result<Addr> {
+        self.end
+            .lock()
+            .unwrap()
+            .as_ref()
+            .map(|end| end.peer_addr())
+            .flatten()
+            .ok_or_else(|| errno!(ENOTCONN, "the socket is not connected"))
+    }
+
+    /// Treat socket with its peer not dropped as connected, regardless of shutdown status.
+    pub fn is_connected(&self) -> bool {
+        if let Some(ref end) = *self.end.lock().unwrap() {
+            end.is_connected()
+        } else {
+            false
+        }
+    }
+
+    pub fn is_blocking(&self) -> bool {
+        self.is_blocking.load(Ordering::SeqCst)
+    }
+
+    pub fn set_non_blocking(&self) {
+        self.is_blocking.store(false, Ordering::SeqCst);
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_non_blocking());
+    }
+
+    pub fn set_blocking(&self) {
+        self.is_blocking.store(true, Ordering::SeqCst);
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_blocking());
+    }
+
+    // TODO: create the corresponding file in the fs
+    pub fn bind(&self, addr: &Addr) -> Result<()> {
+        if !self.addr().is_unnamed() {
+            return_errno!(EINVAL, "the socket is already bound");
+        }
+
+        SERVER_SPACE.add_binder(addr)?;
+
+        *self.addr.write().unwrap() = addr.clone();
+        if let Some(ref end) = *self.end.lock().unwrap() {
+            end.set_addr(&addr);
+        }
+        Ok(())
+    }
+
+    //TODO: add backlog support
+    pub fn listen(&self, backlog: i32) -> Result<()> {
+        if self.addr().is_unnamed() {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+
+        SERVER_SPACE.add_listener(&self.addr())
+    }
+
+    /// A non-blocking accept
+    pub fn accept(&self, flags: FileFlags) -> Result<(Self, Option<Addr>)> {
+        if self.addr().is_unnamed() {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+
+        let sock = SERVER_SPACE.pop_incoming(&self.addr())?;
+
+        if flags.contains(FileFlags::SOCK_NONBLOCK) {
+            sock.set_non_blocking();
+        }
+
+        let addr = sock.peer_addr().ok();
+
+        debug!("the accepted socket is {:?} from {:?}", sock, addr);
+
+        Ok((sock, addr))
+    }
+
+    /// Backlog is temparily not supported so connect will not block
+    pub fn connect(&self, addr: &Addr) -> Result<()> {
+        debug!("connect to {:?}", addr);
+
+        if addr.is_unnamed() {
+            return_errno!(EINVAL, "invalid address");
+        }
+
+        if self.is_connected() {
+            return_errno!(EISCONN, "already connected");
+        }
+
+        if *addr == self.addr() {
+            return_errno!(EINVAL, "self connect is not supported");
+        }
+
```

Comment:
This method is not thread-safe. If the `connect` method is called on a socket concurrently, the socket may connect to multiple different listening sockets, which does not make any sense.